File: .tmp_c4000/batch_04_go/3459_length_of_longest_v_shaped_diagonal_segment.go
// LeetCode 3459 - Length of Longest V-Shaped Diagonal Segment
// https://leetcode.com/problems/length-of-longest-v-shaped-diagonal-segment/

func lenOfVDiagonal(grid [][]int) int {
	m, n := len(grid), len(grid[0])
	dirs := [][2]int{{1, 1}, {1, -1}, {-1, -1}, {-1, 1}}
	// turn clockwise
	nextDir := []int{1, 2, 3, 0}
	ans := 0
	var dfs func(i, j, d, turned, expect int) int
	memo := map[[5]int]int{}
	dfs = func(i, j, d, turned, expect int) int {
		if i < 0 || j < 0 || i >= m || j >= n || grid[i][j] != expect {
			return 0
		}
		key := [5]int{i, j, d, turned, expect}
		if v, ok := memo[key]; ok {
			return v
		}
		ni, nj := i+dirs[d][0], j+dirs[d][1]
		nx := 0
		if expect == 2 {
			nx = 0
		} else {
			nx = 2
		}
		best := 1 + dfs(ni, nj, d, turned, nx)
		if turned == 0 {
			nd := nextDir[d]
			ti, tj := i+dirs[nd][0], j+dirs[nd][1]
			cand := 1 + dfs(ti, tj, nd, 1, nx)
			if cand > best {
				best = cand
			}
		}
		memo[key] = best
		return best
	}
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] != 1 {
				continue
			}
			for d := 0; d < 4; d++ {
				ni, nj := i+dirs[d][0], j+dirs[d][1]
				best := 1 + dfs(ni, nj, d, 0, 2)
				if best > ans {
					ans = best
				}
			}
			if ans < 1 {
				ans = 1
			}
		}
	}
	return ans
}

File: .tmp_rs4013/rust_check17/3459_length_of_longest_v_shaped_diagonal_segment.rs
struct Solution;
// LeetCode 3459 - Length of Longest V-Shaped Diagonal Segment
// https://leetcode.com/problems/length-of-longest-v-shaped-diagonal-segment/

use std::collections::HashMap;

impl Solution {
    pub fn len_of_v_diagonal(grid: Vec<Vec<i32>>) -> i32 {
        let m = grid.len();
        let n = grid[0].len();
        let dirs = [[1, 1], [1, -1], [-1, -1], [-1, 1]];
        let next_dir = [1, 2, 3, 0];
        let mut ans = 0;
        let mut memo: HashMap<(i32, i32, i32, i32, i32), i32> = HashMap::new();
        fn dfs(
            i: i32,
            j: i32,
            d: usize,
            turned: i32,
            expect: i32,
            grid: &[Vec<i32>],
            dirs: &[[i32; 2]; 4],
            next_dir: &[usize; 4],
            memo: &mut HashMap<(i32, i32, i32, i32, i32), i32>,
        ) -> i32 {
            let m = grid.len() as i32;
            let n = grid[0].len() as i32;
            if i < 0 || j < 0 || i >= m || j >= n || grid[i as usize][j as usize] != expect {
                return 0;
            }
            let key = (i, j, d as i32, turned, expect);
            if let Some(&v) = memo.get(&key) {
                return v;
            }
            let ni = i + dirs[d][0];
            let nj = j + dirs[d][1];
            let nx = if expect == 2 { 0 } else { 2 };
            let mut best = 1 + dfs(ni, nj, d, turned, nx, grid, dirs, next_dir, memo);
            if turned == 0 {
                let nd = next_dir[d];
                let ti = i + dirs[nd][0];
                let tj = j + dirs[nd][1];
                let cand = 1 + dfs(ti, tj, nd, 1, nx, grid, dirs, next_dir, memo);
                if cand > best {
                    best = cand;
                }
            }
            memo.insert(key, best);
            best
        }
        for i in 0..m {
            for j in 0..n {
                if grid[i][j] != 1 {
                    continue;
                }
                for d in 0..4 {
                    let ni = i as i32 + dirs[d][0];
                    let nj = j as i32 + dirs[d][1];
                    let best = 1 + dfs(
                        ni, nj, d, 0, 2, &grid, &dirs, &next_dir, &mut memo,
                    );
                    if best > ans {
                        ans = best;
                    }
                }
                if ans < 1 {
                    ans = 1;
                }
            }
        }
        ans
    }
}

fn main() {}
